### ISSUE TITLE
Fix flaky tests (compilation)

### ISSUE THREAD

**Issue by yondonfu** [livepeer/go-livepeer#2145]
This is a thread compiling all known flaky tests that should be fixed.

**Comment by yondonfu** [comment]
Data race in HandlePush()

```
==================
WARNING: DATA RACE
Read at 0x000003084930 by goroutine 132:
  github.com/livepeer/go-livepeer/server.(*LivepeerServer).HandlePush.func4()
      /home/runner/work/go-livepeer/go-livepeer/server/mediaserver.go:954 +0x7c

Previous write at 0x000003084930 by goroutine 84:
  github.com/livepeer/go-livepeer/server.setupServerWithCancel()
      /home/runner/work/go-livepeer/go-livepeer/server/mediaserver_test.go:83 +0x175
  github.com/livepeer/go-livepeer/server.TestPush_ReuseIntmidWithDiffExtmid()
      /home/runner/work/go-livepeer/go-livepeer/server/push_test.go:1469 +0x2fc
  testing.tRunner()
      /opt/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1123 +0x202

Goroutine 132 (running) created at:
  github.com/livepeer/go-livepeer/server.(*LivepeerServer).HandlePush()
      /home/runner/work/go-livepeer/go-livepeer/server/mediaserver.go:952 +0x1b68
  github.com/livepeer/go-livepeer/server.(*LivepeerServer).HandlePush-fm()
      /home/runner/work/go-livepeer/go-livepeer/server/mediaserver.go:774 +0x64
  net/http.HandlerFunc.ServeHTTP()
      /opt/hostedtoolcache/go/1.15.5/x64/src/net/http/server.go:2042 +0x51
  github.com/livepeer/go-livepeer/server.TestPush_ConcurrentSegments.func2()
      /home/runner/work/go-livepeer/go-livepeer/server/push_test.go:1432 +0x38a

Goroutine 84 (running) created at:
  testing.(*T).Run()
      /opt/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1168 +0x5bb
  testing.runTests.func1()
      /opt/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1439 +0xa6
  testing.tRunner()
      /opt/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1123 +0x202
  testing.runTests()
      /opt/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1437 +0x612
  testing.(*M).Run()
      /opt/hostedtoolcache/go/1.15.5/x64/src/testing/testing.go:1345 +0x3b3
  main.main()
      _testmain.go:503 +0x236
==================
E1220 19:48:06.661361   12628 mediaserver.go:992] manifestID=intmid nonce=14955545180663843033 seqNo=0 No sessions available name=0.ts url=http://example.com/live/extmid1/0.ts
E1220 19:48:06.668142   12628 mediaserver.go:992] manifestID=intmid nonce=16845527246296017065 seqNo=0 No sessions available name=0.ts url=http://example.com/live/extmid2/0.ts
--- FAIL: TestPush_ReuseIntmidWithDiffExtmid (0.23s)
    testing.go:1038: race detected during execution of test
E1220 19:48:06.976164   12628 broadcast.go:1213] manifestID=mani nonce=7 seqNo=17 orchSessionID=bar Playlist insertion error nonce=7 manifestID=mani seqNo=17 err="segment already exists"
Could not create StreamContext from binary data
E1220 19:48:07.035252   12628 broadcast.go:511] manifestID=mani nonce=7 seqNo=18 orchSessionID=bar error comparing perceptual hashes from url=https://127.0.0.1:44815/transcoded/segment.ts.phash err="InvalidSignData"
E1220 19:48:07.046457   12628 broadcast.go:1213] manifestID=mani nonce=7 seqNo=18 orchSessionID=bar Playlist insertion error nonce=7 manifestID=mani seqNo=18 err="segment already exists"
FAIL
exit status 1
FAIL	github.com/livepeer/go-livepeer/server	1.908s
```

https://github.com/livepeer/go-livepeer/runs/4586888375?check_suite_focus=true

**Comment by yondonfu** [comment]
```
E1220 19:55:02.909513    7903 selection.go:199] failed to read stake weights for selection err="Stakes error"
--- FAIL: TestMinLSSelector_SelectUnknownSession_RandFreq (0.00s)
    selection_test.go:500: 
        	Error Trace:	selection_test.go:500
        	Error:      	Should not be: []byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x30}
        	Test:       	TestMinLSSelector_SelectUnknownSession_RandFreq
    selection_test.go:505: 
        	Error Trace:	selection_test.go:505
        	Error:      	Not equal: 
        	            	expected: []byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x37}
        	            	actual  : []byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x30}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -2,3 +2,3 @@
        	            	  00000000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
        	            	- 00000010  00 00 00 37                                       |...7|
        	            	+ 00000010  00 00 00 30                                       |...0|
        	            	 }
        	Test:       	TestMinLSSelector_SelectUnknownSession_RandFreq
I1220 19:55:12.385539    7903 webserver.go:1225] Price per pixel set to 1 wei for 1 pixels
FAIL
FAIL	github.com/livepeer/go-livepeer/server	21.794s
```

https://github.com/livepeer/go-livepeer/runs/4586963983?check_suite_focus=true

**Comment by yondonfu** [comment]
```
--- FAIL: TestSubmitSegment_Timeout (0.38s)
    segment_rpc_test.go:2013: 
        	Error Trace:	segment_rpc_test.go:2013
        	Error:      	Not equal: 
        	            	expected: "body timeout: context deadline exceeded"
        	            	actual  : "header timeout: Post \"https://127.0.0.1:46429/segment\": context canceled"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-body timeout: context deadline exceeded
        	            	+header timeout: Post "https://127.0.0.1:46429/segment": context canceled
        	Test:       	TestSubmitSegment_Timeout
I1220 20:08:43.154188    7852 segment_rpc.go:484] orchSessionID=bar Submitting segment bytes=1024 orch=https://127.0.0.1:35257 timeout=8s 
```

https://github.com/livepeer/go-livepeer/runs/4587090305?check_suite_focus=true

**Comment by yondonfu** [comment]
```
--- FAIL: TestSubmitSegment_Timeout (0.46s)
    segment_rpc_test.go:2020: 
        	Error Trace:	segment_rpc_test.go:2020
        	Error:      	Expected nil, but got: &fmt.wrapError{msg:"header timeout: Post \"https://127.0.0.1:44637/segment\": context canceled", err:(*url.Error)(0xc0015cd200)}
        	Test:       	TestSubmitSegment_Timeout
```

https://github.com/livepeer/go-livepeer/runs/4587359191?check_suite_focus=true

**Comment by yondonfu** [comment]
```
    segment_rpc_test.go:2038: 
        	Error Trace:	segment_rpc_test.go:2038
        	Error:      	Expected nil, but got: &fmt.wrapError{msg:"header timeout: Post \"https://127.0.0.1:38727/segment\": context canceled", err:(*url.Error)(0xc001504f30)}
        	Test:       	TestSubmitSegment_Timeout
    segment_rpc_test.go:2046: 
        	Error Trace:	segment_rpc_test.go:2046
        	Error:      	Not equal: 
        	            	expected: "body timeout: context deadline exceeded"
        	            	actual  : "header timeout: Post \"https://127.0.0.1:38727/segment\": context canceled"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-body timeout: context deadline exceeded
        	            	+header timeout: Post "https://127.0.0.1:38727/segment": context canceled
        	Test:       	TestSubmitSegment_Timeout
```

https://github.com/livepeer/go-livepeer/runs/4587531174?check_suite_focus=true

**Comment by yondonfu** [comment]
`go test -run TestSubmitSegment_Timeout` consistently fails when run off of `master` locally with:

```
--- FAIL: TestSubmitSegment_Timeout (0.44s)
    segment_rpc_test.go:1993:
                Error Trace:    segment_rpc_test.go:1993
                Error:          Expected nil, but got: &fmt.wrapError{msg:"header timeout: Post \"https://127.0.0.1:65531/segment\": context canceled", err:(*url.Error)(0xc0005883c0)}
                Test:           TestSubmitSegment_Timeout
FAIL
exit status 1
FAIL    github.com/livepeer/go-livepeer/server  2.080s
```

I am running go1.15.5. This error happens in the same `TestSubmitSegment_Timeout` test function as a few of the other reports listed above.

This particularly test failure is consistently reproducible so I would not classify this is a just a flaky test.

**Comment by darkdarkdragon** [comment]
@yondonfu 
1. data race is around test variables - there is synchronisation between tests incorrectly done, just needs to be fixed
2. TestMinLSSelector_SelectUnknownSession_RandFreq - it is by definition flaky because it relies on random number not hit certain choice - this test fails in 10% of cases
3. TestSubmitSegment_Timeout: I created new virtual machine, installed go1.15.5 - test passes. I was able to fail this test only by stressing CPU to 100% - so I think this test just have too tight timeouts - probably we need to relax restrictions in that test

**Comment by cyberj0g** [comment]
1. @darkdarkdragon could you elaborate on that a bit?
2. fixed
3. fixed

**Comment by darkdarkdragon** [comment]
@cyberj0g 
There is `httpPushResetTimer` variable that is used in tests. It is used in `HandlePush` in goroutine 
```go
	go func() {
		for {
			tick, cancel := httpPushResetTimer()
			select {
			case <-requestEnded:
				cancel()
				return
			case <-tick.Done():
				clog.V(common.VERBOSE).Infof(ctx, "watchdog reset seq=%d dur=%v started=%v", seq, duration, now)
				s.connectionLock.RLock()
				if cxn, exists := s.rtmpConnections[mid]; exists {
					cxn.lastUsed = time.Now()
				}
				s.connectionLock.RUnlock()
			}
		}
	}()
```
So when tests end, this goroutine still runs and reads that variable. In the same time new test starts and writes to this variable - hence the data race.
There is `pushResetWg` variable in tests that tries to fix this issue (using `wgWait` function in tests) - but it is by itself incorrectly used - because if previous test finishes completely (including stopping mentioned goroutine that uses `httpPushResetTimer` variable) - then new tests calls `pushResetWg.Wait` method when wait counter is zero - which is incorrect (from doc: `Note that calls with a positive delta that occur when the counter is zero must happen before a Wait.`)

**Comment by yondonfu** [comment]
Closed by #2161 . Created #2179 to track the remaining data race condition which we're addressing separately.
